_posts/2019-09-15-releasing-an-umbrella-app-with-docker-and-mix-release.md

```diff
@@ -0,0 +1,376 @@
+
+```
+WORKDIR /app/apps/registrar_web
+RUN MIX_ENV=prod mix compile
+RUN npm install --prefix ./assets
+RUN npm run deploy --prefix ./assets
+RUN mix phx.digest
+```
+
+Then we use `mix release` to build our release according to the configuration in the `:releases` key of the `project/0` function in our `mix.exs` file.
+
+```
+WORKDIR /app
+RUN MIX_ENV=prod mix release
+```
+
+This builds our release and places it in `_build/prod/rel/registrar_umbrella`.
+
+Finally, we copy the release into our container and specify that the start script is in `./bin/start`.
+
+Let's talk about that start script now.
+
+## The Start Script
+
+Starting our release is simple. Our `./bin/start` script looks like this:
+
+```bash
+#!/usr/bin/env bash
+
+set -e
+
+echo "Starting app..."
+bin/registrar_umbrella start
+```
+
+At this point, you _might_ be remembering that Distillery provides a "boot hook" feature that allows you to run certain commands/execute some code when the app starts up. You _might_ be wondering how we can accomplish the same goal using `mix release`. How can we, for example, ensure that our migrations run whenever the release starts up? Keep reading to find out!
+
+## Pre-Start Scripts with `rel/env.sh`
+
+The `rel/env.sh` file that was generated by `mix release.init` will run when the release starts. This is where we'll call on our migration script.
+
+Assume we have a module, `Registrar.ReleaseTasks` with a function, `migrate/0` that starts up the application and executes the Ecto migrations:
+
+```elixir
+defmodule Registrar.ReleaseTasks do
+  @moduledoc false
+
+  @start_apps [
+    :crypto,
+    :ssl,
+    :postgrex,
+    :ecto,
+    :ecto_sql
+  ]
+
+  @repos Application.get_env(:registrar, :ecto_repos, [])
+
+  def migrate do
+    start_services()
+    run_migrations()
+    stop_services()
+  end
+
+  defp start_services do
+    IO.puts("Starting dependencies..")
+    # Start apps necessary for executing migrations
+    Enum.each(@start_apps, &Application.ensure_all_started/1)
+
+    # Start the Repo(s) for app
+    IO.puts("Starting repos..")
+
+    # Switch pool_size to 2 for ecto > 3.0
+    Enum.each(@repos, & &1.start_link(pool_size: 2))
+  end
+
+  defp stop_services do
+    IO.puts("Success!")
+    :init.stop()
+  end
+
+  defp run_migrations do
+    Enum.each(@repos, &run_migrations_for/1)
+  end
+end
+```
+
+We can execute this function in our release using `eval`. A call `bin/MY_RELEASE eval` will start up your release and execute whatever function you give as an argument to `eval`. To execute our migration function in our release:
+
+```
+bin/registrar_umbrella eval "Registrar.ReleaseTasks.migrate()"
+```
+
+Recall that we're starting our release in `./bin/start` with that `start` command:
+
+```
+bin/registrar_umbrella start
+```
+
+This will execute the `rel/env.sh` file in turn. This file should contain a script that does the following:
+
+* If the command given to the release was `start`, run the migrations using `eval`.
```

Comment:
On the one hand, since `env.sh` runs when the release is started (for example via the `start` command) this feels like it _could_ be the right place to execute migrations. However, since a call to `eval` _also_ starts the release running, that would execute `env.sh` _again_, hence the need for the `if` condition below. So I'm not sure if running migrations belongs here or if it would be better to put in in `./bin/start` just before the call to `bin/RELEASE_NAME start`. Does anyone have an opinion here? The docs didn't really clear it up for me.

Reply:
I don't have any thoughts. Is there anyone we could reach out to? Or maybe it might be worth putting a note into the blog post about this and solicit feedback?